hedgehog/src/Hedgehog/Internal/Gen.hs

```diff
@@ -1589,16 +1590,40 @@ subsequence xs =
 --   /list./
 --
 shuffle :: MonadGen m => [a] -> m [a]
-shuffle = \case
-  [] ->
-    pure []
-  xs0 -> do
-    n <- integral $ Range.constant 0 (length xs0 - 1)
-    case splitAt n xs0 of
-      (xs, y : ys) ->
-        (y :) <$> shuffle (xs ++ ys)
-      (_, []) ->
-        error "Hedgehog.Gen.shuffle: internal error, split generated empty list"
+-- We shuffle sequences instead of lists to make extracting an arbitrary
+-- element logarithmic instead of linear, and to make length calculation
+-- constant-time instead of linear. We could probably do better, but
+-- this is at least reasonably quick.
+shuffle = fmap toList . shuffleSeq . Seq.fromList
+
+-- | Generates a random permutation of a sequence.
+--
+--   /This shrinks towards the order of the sequence being identical to the input/
+--   /sequence./
+--
+shuffleSeq :: MonadGen m => Seq a -> m (Seq a)
+shuffleSeq xs
+  | null xs = pure Seq.empty
+  | otherwise = do
+    n <- integral $ Range.constant 0 (length xs - 1)
+#if MIN_VERSION_containers(0,5,8)
+    -- Data.Sequence should offer a version of deleteAt that returns the
+    -- deleted element, but it does not currently do so. Lookup followed
+    -- by deletion seems likely faster than splitting and then appending,
+    -- but I haven't actually tested that. It's certainly easier to see
+    -- what's going on.
+    case Seq.lookup n xs of
+      Just y -> (y Seq.<|) <$> shuffleSeq (Seq.deleteAt n xs)
+      Nothing -> error $
+        "Hedgehog.Gen.shuffleSeq: internal error, lookup in empty sequence"
+#else
```

Comment:
![image](https://user-images.githubusercontent.com/287532/71787193-2f0c4080-3015-11ea-99a7-ee11b8ca0e54.png)

The code in `#if MIN_VERSION_containers(0,5,8)` (shown as `Gen.shuffle_fast_1`) is actually faster than the one in the `#else` branch (shown as `Gen.shuffle_fast_2`.)

Do we need to have both? It would be good to avoid CPP where possible.

Reply:
Oh, is it because of `containers                      >= 0.4        && < 0.7`?